src/Servers.jl

```diff
@@ -410,104 +307,78 @@ function handle_connection(f::Function, c::Connection;
     return
 end
 
+"creates an async thread that waits a specified amount of time before closing the connection"
+function check_readtimeout(c, readtimeout, wait_for_timeout)
+    @async while wait_for_timeout[]
+        if inactiveseconds(c) > readtimeout
+            @warn "Connection Timeout: $c"
+            try
+                writeheaders(c.io, Response(408, ["Connection" => "close"]))
+            finally
+                close(c)
+            end
+            break
+        end
+        sleep(8 + rand() * 4)
+    end
+    return
+end
 
 """
-Create a `HTTP.Stream` and parse the Request headers from a `HTTP.Transaction`.
-If there is a parse error, send an error Response.
-Otherwise, execute stream processing function `f`.
+Transaction handler: creates a new Stream for the Transaction, calls startread on it,
+then dispatches the stream to the user-provided handler function. Catches errors on all
+IO operations and closes gracefully if encountered.
 """
-function handle_transaction(f::Function, t::Transaction;
-                            final_transaction::Bool=false,
-                            verbose::Bool=false, kw...)
-
-    request = HTTP.Request()
-    http = Streams.Stream(request, t)
+function handle(h, t::Transaction, last::Bool=false)
+    request = Request()
+    http = Stream(request, t)
 
     try
         startread(http)
     catch e
-        # @show typeof(e)
-        # @show fieldnames(e)
         if e isa EOFError && isempty(request.method)
             return
-# FIXME https://github.com/JuliaWeb/HTTP.jl/pull/178#pullrequestreview-92547066
-#        elseif !isopen(http)
-#            @warn "Connection closed"
-#            return
-        elseif e isa HTTP.ParseError
-            @error e
+        elseif e isa ParseError
             status = e.code == :HEADER_SIZE_EXCEEDS_LIMIT  ? 413 : 400
             write(t, Response(status, body = string(e.code)))
             close(t)
             return
+        elseif e isa Base.IOError && e.code == -54
+            # read: connection reset by peer (ECONNRESET)
+            return
         else
             rethrow(e)
         end
     end
 
-    if verbose
-        @info http.message
-    end
-
-    response = request.response
-    response.status = 200
-    if final_transaction || hasheader(request, "Connection", "close")
-        setheader(response, "Connection" => "close")
-    end
-
-    @async try
```

Comment:
Does removing this `@async` have the effect of removing request pipelining?

Reply:
I meant to ask you about it. Do we have tests around request pipelining? Removing it didn't cause any tests to fail. Honestly I don't really understand how request pipelining is supposed to work (i.e. how the requests are kept separate & uniquely identified server-side).

Reply:
> I don't really understand how request pipelining is supposed to work

The pipelining mechanism is build into `ConnectionPool.Transaction`.

> how the requests are kept separate & uniquely identified server-side

That is handled by the `Transaction` sequence number: https://github.com/JuliaWeb/HTTP.jl/blob/master/src/ConnectionPool.jl#L248-L250

There are test cases that check request interleaving with different pipeline depth limits here: 
https://github.com/JuliaWeb/HTTP.jl/blob/master/test/loopback.jl#L297-L412
These tests are driven by the client side interface. It would be good to implement similar tests for the server side, however the basic pipelining mechanism is the same regardless of client/server.

Reply:
(I have a pressing time commitment on a different project for the next few of days. I'll try to find a few hours when I can to help with this PR, but I won't have a chance to do a proper review of the implementation details until later in the week.)

Reply:
Note I just found `autocannon -p` which does load testing with pipelining. https://github.com/JuliaWeb/HttpServer.jl/issues/142#issuecomment-426465879

Reply:
Here is a simple example of the effects of pipelining.
```julia
using HTTP ; HTTP.listen("127.0.0.1", 8088; verbose=true) do http::HTTP.Stream
           body = read(http); closeread(http)
           sleep(2) # Time taken to process request.
           startwrite(http); write(http, body)
       end
```

```julia
@time @sync for i in 1:4
    @async HTTP.get("http://127.0.0.1:8088/$i"; connection_limit=1)
end
```

Without the `@async` around `try handle_stream` the test takes about 8 seconds. With the `@async` it takes about 2.

Or, using `autocannon` without `@async`:
```
$ autocannon http://127.0.0.1:8088 -c1 -p8
Running 10s test @ http://127.0.0.1:8088
1 connections with 8 pipelining factor

Stat    2.5%    50%     97.5%   99%     Avg        Stdev   Max
Latency 2002 ms 2003 ms 2016 ms 2016 ms 2006.25 ms 5.68 ms 2016.73 ms

Stat      1%  2.5% 50% 97.5% Avg    Stdev  Min
Req/Sec   0   0    0   1     0.4    0.49   1
Bytes/Sec 0 B 0 B  0 B 52 B  20.8 B 25.5 B 52 B

Req/Bytes counts sampled once per second.

4 requests in 10s, 208 B read
```

and with `@async` the latency is still 2000ms but the throughput is 8 times higher:
```
$ autocannon http://127.0.0.1:8088 -c1 -p8
Running 10s test @ http://127.0.0.1:8088
1 connections with 8 pipelining factor

Stat    2.5% 50%  97.5%   99%     Avg       Stdev     Max
Latency 0 ms 0 ms 2018 ms 2018 ms 250.97 ms 664.01 ms 2018.31 ms

Stat      1%  2.5% 50% 97.5% Avg   Stdev Min
Req/Sec   0   0    0   8     3.2   3.92  8
Bytes/Sec 0 B 0 B  0 B 416 B 166 B 204 B 416 B

Req/Bytes counts sampled once per second.

32 requests in 10s, 1.66 kB read
```

Reply:
Also noticed we probably should do this so that the read half of the transaction is freed up before we go into the request processing function.
```patch
--- a/src/Servers.jl
+++ b/src/Servers.jl
@@ -504,12 +504,13 @@ end
 """
 Execute Request processing function `f(::HTTP.Request) -> HTTP.Response`.
 """
 function handle_request(f::Function, http::Stream)
     request::HTTP.Request = http.message
     request.body = read(http)
+    closeread(http)
     request.response::HTTP.Response = f(request)
     request.response.request = request
     startwrite(http)
     write(http, request.response.body)
     return
 end
```